webapp/src/js/services/privacy-policies.js

```diff
@@ -0,0 +1,148 @@
+    };
+
+    const checkAcceptanceLog = (languageCode, privacyPolicies, userSettings) => {
+      if (!userSettings.privacy_policy_acceptance_log || !userSettings.privacy_policy_acceptance_log.length) {
+        return false;
+      }
+
+      const attachmentName = privacyPolicies.privacy_policies[languageCode];
+      const attachment = privacyPolicies._attachments[attachmentName];
+      const entry = userSettings.privacy_policy_acceptance_log.find(entry => {
+        return entry.language === languageCode && entry.digest === attachment.digest;
+      });
+
+      return !!entry;
+    };
+
+    /**
+     * Checks remote and local meta databases to verify whether the current privacy policy has been accepted
+     * Checking remote is required after an initial replication, as meta db sync begins late after app bootstrapping,
+     * so the chances are high we have not downloaded the doc locally yet.
+     * If any of the logs show the current privacy policy was accepted, returns { accepted: true }
+     */
+    const hasAccepted = () => {
+      return $q
+        .all([
+          Language(),
+          getPrivacyPolicies(),
+          UserSettings(),
+        ])
+        .then(([ languageCode, privacyPolicies, userSettings ]) => {
+          if (!policyForLanguageExists(languageCode, privacyPolicies)) {
+            return { privacyPolicy: false, accepted: true };
+          }
+
+          if (checkAcceptanceLog(languageCode, privacyPolicies, userSettings)) {
+            return { privacyPolicy: true, accepted: true };
+          }
+
+          return { privacyPolicy: true, accepted: false };
+        });
+    };
+
+    // atob doesn't handle unicode characters
+    // stolen from StackOverflow
+    const decodeUnicode = string => {
+      // Going backwards: from byte stream, to percent-encoding, to original string.
+      const unicodeCharArray = atob(string)
+        .split('')
+        .map(char => '%' + ('00' + char.charCodeAt(0).toString(16)).slice(-2))
+        .join('');
+      return decodeURIComponent(unicodeCharArray);
+    };
+
+    const getPrivacyPolicy = () => {
+      return $q
+        .all([
+          Language(),
+          getPrivacyPolicies(true),
+        ])
+        .then(([languageCode, privacyPolicies]) => {
+          const attachmentName = privacyPolicies && privacyPolicies.privacy_policies[languageCode];
+          if (!attachmentName || !privacyPolicies._attachments[attachmentName]) {
+            return false;
+          }
+
+          const attachment = privacyPolicies._attachments[attachmentName];
+          if (attachment.content_type !== ACCEPTED_CONTENT_TYPE) {
+            return false;
+          }
+
+          const encodedContent = attachment.data;
+          const decodedContent = decodeUnicode(encodedContent);
+          return {
+            language: languageCode,
+            digest: attachment.digest,
+            html: $sce.trustAsHtml(decodedContent),
```

Comment:
I don't think we should trust this content, because we don't have any control over it. As far as I'm concerned app builders should only be uploading markup for formatting, not for doing anything like scripting. What do you think?

Reply:
I believe Angular won't display it unless we trust it.

Reply:
You want to use the $sanitize service to allow safe tags to be used: https://docs.angularjs.org/api/ngSanitize/service/$sanitize

I thought this happened automatically if you didn't call `$sce.trustAsHtml` but maybe there's something you need to do to get this to work?

Reply:
Updated to use $sanitize